File: src/csv_handler.py
"""CSV import/export for spin data."""

import io
from datetime import datetime

import pandas as pd

from src.database import get_connection, get_or_create_dealer, bulk_insert_spins
from src.models import BALL_SIZES, SPIN_SPEEDS, WHEEL_SPEEDS, BALL_DIRECTIONS

REQUIRED_COLUMNS = ["dealer_name", "result_number"]
OPTIONAL_COLUMNS = ["ball_size", "spin_speed", "wheel_speed", "ball_direction",
                     "session_tag", "notes", "recorded_at"]

VALID_VALUES = {
    "ball_size": set(BALL_SIZES) | {None, "", "nan"},
    "spin_speed": set(SPIN_SPEEDS) | {None, "", "nan"},
    "wheel_speed": set(WHEEL_SPEEDS) | {None, "", "nan"},
    "ball_direction": set(BALL_DIRECTIONS) | {None, "", "nan"},
}


def validate_csv(uploaded_file) -> tuple[bool, list[str], pd.DataFrame | None]:
    """Validate an uploaded CSV file.

    Returns (is_valid, error_messages, preview_dataframe).
    """
    errors = []
    try:
        df = pd.read_csv(uploaded_file)
    except Exception as e:
        return False, [f"Could not parse CSV: {e}"], None

    # Check required columns
    for col in REQUIRED_COLUMNS:
        if col not in df.columns:
            errors.append(f"Missing required column: '{col}'")

    if errors:
        return False, errors, df.head()

    # Validate result_number range
    invalid_nums = df[~df["result_number"].between(0, 36)]
    if not invalid_nums.empty:
        errors.append(
            f"{len(invalid_nums)} rows have result_number outside 0-36 "
            f"(rows: {invalid_nums.index.tolist()[:5]})"
        )

    # Validate enum fields
    for col, valid in VALID_VALUES.items():
        if col in df.columns:
            df[col] = df[col].astype(str).replace("nan", "")
            invalid = df[~df[col].isin(valid) & (df[col] != "")]
            if not invalid.empty:
                bad_vals = invalid[col].unique()[:5].tolist()
                errors.append(f"Invalid values in '{col}': {bad_vals}")

    is_valid = len(errors) == 0
    return is_valid, errors, df.head(10)


def import_csv(uploaded_file) -> tuple[int, int, list[str]]:
    """Import spins from CSV. Auto-creates dealers if needed.

    Returns (n_imported, n_skipped, errors).
    """
    df = pd.read_csv(uploaded_file)
    errors = []
    n_imported = 0
    n_skipped = 0

    # Clean up enum fields
    for col in VALID_VALUES:
        if col in df.columns:
            df[col] = df[col].astype(str).replace("nan", "")
            df[col] = df[col].apply(lambda x: x if x in VALID_VALUES[col] and x != "" else None)

    # Resolve dealer IDs
    dealer_ids = {}
    for name in df["dealer_name"].unique():
        name_str = str(name).strip()
        if name_str and name_str != "nan":
            dealer_ids[name_str] = get_or_create_dealer(name_str)

    # Build insert DataFrame
    rows_to_insert = []
    for idx, row in df.iterrows():
        dealer_name = str(row["dealer_name"]).strip()
        if dealer_name not in dealer_ids:
            n_skipped += 1
            errors.append(f"Row {idx}: empty dealer name, skipped")
            continue

        result_num = row["result_number"]
        if not (0 <= int(result_num) <= 36):
            n_skipped += 1
            errors.append(f"Row {idx}: invalid result_number {result_num}, skipped")
            continue

        rows_to_insert.append({
            "dealer_id": dealer_ids[dealer_name],
            "result_number": int(result_num),
            "ball_size": row.get("ball_size") if pd.notna(row.get("ball_size")) else None,
            "spin_speed": row.get("spin_speed") if pd.notna(row.get("spin_speed")) else None,
            "wheel_speed": row.get("wheel_speed") if pd.notna(row.get("wheel_speed")) else None,
            "ball_direction": row.get("ball_direction") if pd.notna(row.get("ball_direction")) else None,
            "session_tag": str(row.get("session_tag", "")) if pd.notna(row.get("session_tag")) else "",
            "notes": str(row.get("notes", "")) if pd.notna(row.get("notes")) else "",
            "recorded_at": str(row.get("recorded_at", "")) if pd.notna(row.get("recorded_at")) else datetime.now().isoformat(),
        })

    if rows_to_insert:
        insert_df = pd.DataFrame(rows_to_insert)
        bulk_insert_spins(insert_df)
        n_imported = len(rows_to_insert)

    return n_imported, n_skipped, errors


def export_csv(dealer_id: int = None) -> str:
    """Export spins to CSV string. If dealer_id is None, exports all."""
    conn = get_connection()

    if dealer_id:
        query = """
            SELECT d.name as dealer_name, s.result_number, s.ball_size, s.spin_speed,
                   s.wheel_speed, s.ball_direction, s.session_tag, s.notes, s.recorded_at
            FROM spins s JOIN dealers d ON s.dealer_id = d.id
            WHERE s.dealer_id = ?
            ORDER BY s.recorded_at
        """
        df = pd.read_sql_query(query, conn, params=(dealer_id,))
    else:
        query = """
            SELECT d.name as dealer_name, s.result_number, s.ball_size, s.spin_speed,
                   s.wheel_speed, s.ball_direction, s.session_tag, s.notes, s.recorded_at
            FROM spins s JOIN dealers d ON s.dealer_id = d.id
            ORDER BY d.name, s.recorded_at
        """
        df = pd.read_sql_query(query, conn)

    return df.to_csv(index=False)

File: src/data/csv-handler.js
/** CSV import/export for spin data. */

import { db } from '../db/schema.js';
import { getOrCreateDealer } from '../db/dealers.js';
import { bulkInsertSpins } from '../db/spins.js';

const BALL_SIZES = new Set(['small', 'medium', 'large']);
const SPIN_SPEEDS = new Set(['slow', 'medium', 'fast']);
const WHEEL_SPEEDS = new Set(['slow', 'medium', 'fast']);
const BALL_DIRS = new Set(['cw', 'ccw']);

function parseCSV(text) {
  const lines = text.trim().split('\n');
  if (lines.length < 2) return { headers: [], rows: [] };

  const headers = lines[0].split(',').map(h => h.trim());
  const rows = [];

  for (let i = 1; i < lines.length; i++) {
    const values = lines[i].split(',').map(v => v.trim());
    const row = {};
    headers.forEach((h, j) => { row[h] = values[j] || ''; });
    rows.push(row);
  }

  return { headers, rows };
}

export function validateCSV(text) {
  const { headers, rows } = parseCSV(text);
  const errors = [];

  if (!headers.includes('dealer_name')) errors.push("Missing required column: 'dealer_name'");
  if (!headers.includes('result_number')) errors.push("Missing required column: 'result_number'");
  if (errors.length > 0) return { valid: false, errors, preview: rows.slice(0, 10) };

  let invalidNums = 0;
  for (let i = 0; i < rows.length; i++) {
    const num = parseInt(rows[i].result_number);
    if (isNaN(num) || num < 0 || num > 36) invalidNums++;
  }
  if (invalidNums > 0) errors.push(`${invalidNums} rows have result_number outside 0-36`);

  for (const [col, valid] of [['ball_size', BALL_SIZES], ['spin_speed', SPIN_SPEEDS], ['wheel_speed', WHEEL_SPEEDS], ['ball_direction', BALL_DIRS]]) {
    if (headers.includes(col)) {
      const bad = new Set();
      for (const row of rows) {
        const val = row[col];
        if (val && val !== '' && !valid.has(val)) bad.add(val);
      }
      if (bad.size > 0) errors.push(`Invalid values in '${col}': ${[...bad].slice(0, 5).join(', ')}`);
    }
  }

  return { valid: errors.length === 0, errors, preview: rows.slice(0, 10) };
}

export async function importCSV(text) {
  const { headers, rows } = parseCSV(text);
  let imported = 0, skipped = 0;
  const errors = [];
  const spinsToInsert = [];

  for (let i = 0; i < rows.length; i++) {
    const row = rows[i];
    const dealerName = row.dealer_name?.trim();
    if (!dealerName) { skipped++; errors.push(`Row ${i + 1}: empty dealer name`); continue; }

    const num = parseInt(row.result_number);
    if (isNaN(num) || num < 0 || num > 36) { skipped++; errors.push(`Row ${i + 1}: invalid number`); continue; }

    const dealerId = await getOrCreateDealer(dealerName);
    spinsToInsert.push({
      dealerId,
      resultNumber: num,
      ballSize: BALL_SIZES.has(row.ball_size) ? row.ball_size : null,
      spinSpeed: SPIN_SPEEDS.has(row.spin_speed) ? row.spin_speed : null,
      wheelSpeed: WHEEL_SPEEDS.has(row.wheel_speed) ? row.wheel_speed : null,
      ballDirection: BALL_DIRS.has(row.ball_direction) ? row.ball_direction : null,
      sessionTag: row.session_tag || '',
      notes: row.notes || '',
      recordedAt: row.recorded_at || new Date().toISOString(),
    });
  }

  if (spinsToInsert.length > 0) {
    await bulkInsertSpins(spinsToInsert);
    imported = spinsToInsert.length;
  }

  return { imported, skipped, errors };
}

export async function exportCSV(dealerId = null) {
  let spins;
  if (dealerId) {
    spins = await db.spins.where('dealerId').equals(dealerId).toArray();
  } else {
    spins = await db.spins.toArray();
  }

  const dealers = await db.dealers.toArray();
  const dealerMap = new Map(dealers.map(d => [d.id, d.name]));

  const headers = 'dealer_name,result_number,ball_size,spin_speed,wheel_speed,ball_direction,session_tag,notes,recorded_at';
  const rows = spins.map(s =>
    [dealerMap.get(s.dealerId) || '', s.resultNumber, s.ballSize || '', s.spinSpeed || '',
     s.wheelSpeed || '', s.ballDirection || '', s.sessionTag || '', s.notes || '', s.recordedAt || ''].join(',')
  );

  return [headers, ...rows].join('\n');
}
